### PR TITLE
Changing blob to use arc pointers in order to limit wasteful copying …

### LINKED ISSUE #8801: Make Blob store an Arc<Vec<u8>>
The point of Blob values is that they're immutable buffers, and sometimes they're even slices of immutable buffers, and they're often used as sources of data that will be sent between threads internally. We should support this properly by making the underlying buffer threadsafe, allowing us to pass around references to it rather than cloning it. This will allow us to:
* remove the `read_out_buffer` method, and rename `clone_bytes` to `get_bytes`
* make Blob slices (ie. the result of calling `Blob::Slice`) share the original buffer, by storing the offset and length of the slice in the new Blob object

Code: components/script/dom/blob.rs

### PR DESCRIPTION
**Pull request by craftytrickster**
Fixes #8801.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8860)

### CODE REVIEW

File: components/script/dom/blob.rs [hunk 1]
```diff
@@ -70,23 +94,20 @@ impl Blob {
...
-        send.send(self.bytes.clone().unwrap_or(vec![])).unwrap();
-    }
+    pub fn get_bytes(&self) -> Vec<u8> {
+        let start = self.bytes_start.unwrap_or(0);
+        let end = self.bytes_end.unwrap_or(self.bytes.len());
 
-    // simpler to use version of read_out_buffer
-    pub fn clone_bytes(&self) -> Vec<u8> {
-        self.bytes.clone().unwrap_or(vec![])
+        self.bytes[start..end].to_vec()
```
Comment (SimonSapin) [inline components/script/dom/blob.rs]: This (`.to_vec()`) still copies bytes. It looks like this method is used to send `Vec<u8>` messages, so it probably can’t simply return `&[u8]` instead.

To reduce copying, the `Vec<u8>` messages should probably be changed to some reference counted thing. Maybe tendrils? https://github.com/servo/tendril

CC @jdm

Comment (jdm) [inline components/script/dom/blob.rs]: I don't know much about tendril and it doesn't really have any good examples of usage afaict. I'm not sure I see any options here; the websocket library accepts a `Vec<u8>`, so I believe we're going to have to copy at some point here.

Comment (jdm) [inline components/script/dom/blob.rs]: Hmm, I'm wrong - the library does accept `&[u8]`, so I think we should make this return `&[u8]` instead.

**Comment by jdm** [comment]
So looking at the code in filereader.rs, I see we'll need another change to avoid copying while dealing with slices appropriately. I propose we extract the buffer and slice data into a separate structure (`BlobSlice`) that can be sent between threads, and has a method that returns the actual slice (`-> &[u8]`). This struct can be the return value of `get_bytes()` on Blob. Does that make sense?

**Comment by craftytrickster** [comment]
@jdm that does make sense, I'll see what I can do.

**Comment by craftytrickster** [comment]
I pushed a new test commit to show an approach to avoid using `to_vec` in the `get_bytes` method.

However, I do not know if I like the decision to split of the DataSlice/BlobSlice into its own struct, it does not feel very ergonomic to me. I'd prefer to keep those fields in the Blob struct. I guess you can look at the latest commit to see if you agree or not. Also, splitting up the struct caused me to get compiler errors on DataSlice unless I implemented `heap_size_of_children` and `trace`. I was not sure how to remove these errors, so for the prototype I just kept empty implementations.

**Comment by jdm** [comment]
For what it's worth, that looks like what I was expecting :) I agree that it would be nice to keep the representation of the data within the containing blob, but our current design forbids us from being able to access the internals of JS-owned values (such as Blobs) on other threads.

**Comment by nox** [comment]
@jdm What if we had something like:

```rust
pub enum BlobData {
    Owned(Vec<u8>),
    Borrowed(JS<Blob>, usize, usize)
}
```

This would allow us to actually track the memory usage of the blobs (by making BlobData's HeapSizeOf implementation return 0 for Borrowed values).

**Comment by jdm** [comment]
@nox I'm not sure that really helps if the original blob is inaccessible and only the slice is holding it alive.

**Comment by Ms2ger** [comment]
Another thing to keep in mind is that you can `postMessage()` a `Blob` to a worker.

**Comment by nox** [comment]
Oh you are completely right, silly me.

**Comment by KiChjang** [comment]
@craftytrickster Ping?

**Comment by craftytrickster** [comment]
Hi @KiChjang, I am still waiting for any feedback on what needs to be done here. If you have any comments, please let me know

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8860#-:-K7HGDWZlM6zUNJ43HAi:-295904643)*

**Comment by craftytrickster** [comment]
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/blob.rs, line 101 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8860#-K4se3yUipcTSqF631Yj-r1-101:-K7HHdu5bQ9ICfTBIYuq:-417417542) ([raw file](https://github.com/servo/servo/blob/82c9af0a687d9402e8a8ca27dfc7579b6183ce87/components/script/dom/blob.rs#L101)):*
I am no longer doing to_vec and copying the bytes, I am simply passing a reference to the datsaslice now

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8860)*

**Comment by jdm** [comment]
Sorry this fell through the cracks!

**Comment by jdm** [comment]
This looks like what I expected! There are a couple changes we can make to improve clarity, and I think we might be able to avoid duplicating the blob contents in the websocket code, though.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 2 of 5 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/dom/blob.rs, line 106 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8860#-K7I8c3I_XSpvSSmvjdO:-K7I8c3I_XSpvSSmvjdP:-1080798717) ([raw file](https://github.com/servo/servo/blob/4f33b07b60342a3e49d2e502a57a52747823ced8/components/script/dom/blob.rs#L106)):*
On second thought, let's extend the `new` constructor to accept `bytes_start` and `bytes_end` instead of duplicating this.

---

*[components/script/dom/filereader.rs, line 359 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8860#-K7I95x4wGQ9E2U4adlM:-K7I95x4wGQ9E2U4adlN:-266572298) ([raw file](https://github.com/servo/servo/blob/4f33b07b60342a3e49d2e502a57a52747823ced8/components/script/dom/filereader.rs#L359)):*
We don't need a sender and receiver any longer. We can store the result of `blob.get_data().clone()` in a variable and pass that to perform_annotated_read_operation.

---

*[components/script/dom/filereader.rs, line 410 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8860#-K7I9NElEC1sy9b6CIzn:-K7I9NEmSLdS8Ex1khoY:-1366579396) ([raw file](https://github.com/servo/servo/blob/4f33b07b60342a3e49d2e502a57a52747823ced8/components/script/dom/filereader.rs#L410)):*
`blob_contents` should be a `DataSlice` value instead of a Receiver.

---

*[components/script/dom/websocket.rs, line 389 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8860#-K7I9ZFiOBqo4llPOJwE:-K7I9ZFiOBqo4llPOJwF:1173216477) ([raw file](https://github.com/servo/servo/blob/4f33b07b60342a3e49d2e502a57a52747823ced8/components/script/dom/websocket.rs#L389)):*
Can we get rid of the `to_vec` here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8860#-:-K7I9wu1uMwStWXeueIl:-1583556618)*

**Comment by craftytrickster** [comment]
@jdm thank you for the notes, I will take a look at them when I have time tonight and make the appropriate changes

**Comment by craftytrickster** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/dom/blob.rs, line 106 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8860#-K7I8c3I_XSpvSSmvjdO:-K7OKlASJoCSzmyDzdPI:1059832497) ([raw file](https://github.com/servo/servo/blob/4f33b07b60342a3e49d2e502a57a52747823ced8/components/script/dom/blob.rs#L106)):*
Originally, I tried that, but I was having issues compiling with the use of "box". It seems to work fine now though, so I was able to remove the duplicated function body.

However, since new implies a fresh object, it seems awkward to have to specified None, None, every single time in the constructor. I think we should still keep a separated new_sliced function to avoid this.

If you do not agree, I can simply converge the new & new_sliced together. 

---

*[components/script/dom/websocket.rs, line 389 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8860#-K7I9ZFiOBqo4llPOJwE:-K7OU65fjzFAhiPPzOXF:556558897) ([raw file](https://github.com/servo/servo/blob/4f33b07b60342a3e49d2e502a57a52747823ced8/components/script/dom/websocket.rs#L389)):*
This is a bit complicated, because we would have to change the net_traits lib.rs file that contains the MessageData enum. It would have to be changed to Blob instead of a Vec<u8> in order to avoid using to_vec. 

The problem is that we cannot import (script) Blob there, because it would create a circular dependency. I am not sure what you would like to do here. Should we move blob to another crate?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8860)*

**Comment by jdm** [comment]
@bors-servo: r+
Thanks for doing this work!

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/websocket.rs, line 389 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8860#-K7I9ZFiOBqo4llPOJwE:-K7PyZUUR-OfFulHSy7l:1979191221) ([raw file](https://github.com/servo/servo/blob/4f33b07b60342a3e49d2e502a57a52747823ced8/components/script/dom/websocket.rs#L389)):*
Oh right, I forgot that this became more complicated recently. Nothing we can do here!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8860#-:-K7Pyid1yvvEdZSuK66L:1436933519)*

**Comment by KiChjang** [comment]
@bors-servo retry
* #9163

**Pull request merged**